README.md
# Backbeam server

[![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/backbeam/backbeam-server?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge) [![Coverage Status](https://img.shields.io/coveralls/backbeam/backbeam-server.svg)](https://coveralls.io/r/backbeam/backbeam-server?branch=master) [![Build Status](https://travis-ci.org/backbeam/backbeam-server.svg?branch=master)](https://travis-ci.org/backbeam/backbeam-server)

Backbeam-server is a compatible implementation of [Backbeam.io](http://backbeam.io), a cloud backend-as-a-service. This project is maintained by the creators of Backbeam and this project will pass similar tests to keep its compatibility.

## Introduction

### Migrating

If you are upgrading from 0.5.x to 0.6.x, please see the CHANGELOG

### Why an Open Source version?

We want our customers to be happy and one of the things that most worry our customers is the vendor-lockin problem. Providing an Open Source implementation solves this problem.

Another important reason is flexibility. Having the opportunity to host your own backbeam-compatible stack you can do many things such as:

- You can install `backbeam-server` in your Continuous Integration server for testing purpouses
- If you are developing a new SDK for Backbeam.io you can test it locally with `backbeam-server`.
- Soon you will be able to dump your data from Backbeam.io to your machines and import the data to a local instance to inspect the data carefully for: machine learning, data mining, reporting, pattern-finding, etc.

### How does it differ from backbeam.io?

- With `backbeam-server` you will have to implement your infrastructure to support scalability, backups, server configuration, security protection, etc.
- backbeam.io supports code versioning itself. With `backbeam-server` you just need to use your favorite version control tool.

## Usage

* Fist step: install `backbeam-server`: `npm install backbeam-server -g`
* Create a directory where you want to start a project.
* Navigate to that directory with the command line.
* Run `backbeam create`. This will generate the basic structure
* Check the database configuration in the generated `config.json` file.
* Check that `mysql` is running
* Run `backbeam start`.
* Browse `http://localhost:3000` in your browser and you will see an example web controller running.
* Browse `http://localhost:3000/admin` in your browser and you will see the admin panel.
* Change controllers, assets, etc. You don't need to restart backbeam, just refresh your browser.
* If you change the data model in `config.json` you need to run: `backbeam migrate` with the server running or you can go to the admin panel to refresh the configuration (`Configuration → Schmea changes` and then `Reload configuration`). That will show you the required SQL commands to update your database schema. You can run them by clicking on `Run SQL commands`.

## What does it support?

Both backbeam.io and `backbeam-server` provide the following features

- Database
- REST API (with Open Source SDKs for iOS, Android and JavaScript)
- Users authentication (email+password, Twitter, Facebook or Google+)
- Push notifications for iOS and Android
- Server-side code execution in JavaScript
- Full-featured MVC web framework
- A real-time API

## Roadmap

- Almost full `bql` support on top of `mysql`
- Implementation of admin panel
- Implementation of all features (users authentication, push notifications,...)
- Finish implementation of `bql` on top of `mysql`
- Custom `bql` functionlity for `mysql`
- Plugin system


CHANGELOG.md
Changelog
=========

**Version 0.6.0 - Apr 16, 2015**

## Configuration changes

Data model is now defined in `model.json` instead of in `config.json`. To migrate from 0.5.x just move the contents of your `model` section from `config.json` to `model.json`. So your `model.json` file looks like this:

```json
{
  "entities": [ ... ]
}
```

Now you can also load different configuration files depending on the `NODE_ENV` variable. For example if `NODE_ENV=production` you need a `config-production.json` file. If `NODE_ENV` is not set or blank then `config.json` is used.

Also, there's different behavior depending on your `NODE_ENV` variable. If it starts with `prod` then files are not hot-reloaded and templates are cached.

## Changes in controllers

In `routes.json` you can now specify an `action` to any controller. If you don't specify one then `run` is supposed. For example:

```json
{
  "method": "GET",
  "path": "/",
  "file": "home.js",
  "action": "run"
}
```

Then your controller should look like the following:

```javascript
exports.run = function(backbeam, request, response, libs, logger) {
  // your code
}
```

This has two main benefits:

* No global variables used which is good for many reasons
* You can specify more than one controller in each file
* We can add more parameters in the future without having to worry about naming collissions
* You can name your parameters as you wish. Example: `req` instead of `request`
* You can use `return` easily in your controllers. Such as `if (something) return response.send('Hello')`

The new `libs` object has a `require` method that let you require files inside the `libs` folder. For example `var filters = libs.require('filters')`.

The `logger` object let's you log information to the console with `logger.log('Hello world')`. This method will log the given message besides additional information such as a formatted date, the HTTP method, request path and the source file of the invoked controller. There are also `warn` and `error` methods available.


cli/cli-create.js
var fs = require('fs')
var path = require('path')
var mkdirp = require('mkdirp')
var txain = require('txain')
var colors = require('colors')
var program = require('commander')
var version = require('../package.json').version
var crypto = require('crypto')
var _ = require('underscore')
var multiline = require('multiline')

exports.run = function(argv, callback) {
  program
    .version(version)
    .option('--directory [<s>]', 'Directory where the project has to be created', process.cwd())
    .parse(argv)

  console.log('Creating project at %s', program.directory)
  var projectName = path.basename(program.directory)

  txain(function(callback) {
    mkdirp(program.directory, callback)
  })
  .then(function(callback) {
    // configuration
    var configFile = path.join(program.directory, 'config.json')
    var exists = fs.existsSync(configFile)
    if (exists) return callback()
    var data = {
      project: {
        name: projectName,
      },
      db: {
        manager: 'sql',
        host: 'localhost',
        port: 3306,
        user: 'root',
        pass: '',
        database: projectName,
      },
      api: {
        keys: {}
      },
      email: {
        transport: {
          service: 'stub',
        },
        from: 'user@example.com',
        inline: true,
      },
      authentication: {
        sessionKey: crypto.randomBytes(32).toString('hex'),
        email: {
          confirmation: 'optional',
        }
      }
    }
    data.api.keys[randomToken(16)] = randomToken(32)
    fs.writeFile(configFile, JSON.stringify(data, null, 2), 'utf8', callback)
  })
  .then(function(callback) {
    // test configuration
    var configFile = path.join(program.directory, 'config-test.json')
    var exists = fs.existsSync(configFile)
    if (exists) return callback()
    var data = {
      project: {
        name: projectName,
      },
      db: {
        manager: 'sql',
        host: 'localhost',
        port: 3306,
        user: 'root',
        pass: '',
        database: projectName,
      },
      api: {
        keys: {}
      },
      email: {
        transport: {
          service: 'stub',
        },
        from: 'user@example.com',
        inline: true,
      },
      authentication: {
        sessionKey: crypto.randomBytes(32).toString('hex'),
        email: {
          confirmation: 'optional',
        }
      }
    }
    data.api.keys[randomToken(16)] = randomToken(32)
    fs.writeFile(configFile, JSON.stringify(data, null, 2), 'utf8', callback)
  })
  .then(function(callback) {
    var modelFile = path.join(program.directory, 'model.json')
    var exists = fs.existsSync(modelFile)
    if (exists) return callback()
    var model = {
      entities: [
        {
          id: 'item',
          fields: [
            {
              id: 'name',
              type: 'text',
              mandatory: true,
            }
          ]
        }
      ]
    }
    fs.writeFile(modelFile, JSON.stringify(model, null, 2), 'utf8', callback)
  })
  .then(function(callback) {
    // directories
    var dirs = [
      'push',
      'files',
      'web/assets',
      'web/views',
      'web/controllers',
      'web/libs',
      'email_templates',
      'test',
    ]
    dirs = dirs.map(function(dir) {
      return path.join(program.directory, dir)
    })
    txain(dirs).each(mkdirp).end(callback)
  })
  .then(function(callback) {
    callback(null, _.keys(emailTemplates))
  })
  .each(function(type, callback) {
    var template = emailTemplates[type]
    var fullpath = path.join(program.directory, 'email_templates', type)
    txain(mkdirp, fullpath)
    .then(function(callback) {
      callback(null, _.keys(template))
    })
    .each(function(part, callback) {
      var source = template[part]
      fs.writeFile(path.join(fullpath, type+'.'+part), source, 'utf8', callback)
    })
    .end(callback)
  })
  .then(function(callback) {
    // routes
    var routesFile = path.join(program.directory, 'web/controllers/routes.json')
    var exists = fs.existsSync(routesFile)
    if (exists) return callback()
    var data = [
      {
        method: 'GET',
        path: '/',
        file: 'home.js',
        action: 'run',
      }
    ]
    fs.writeFile(routesFile, JSON.stringify(data, null, 2), 'utf8', callback)
  })
  .then(function(callback) {
    // home controller
    var controllerFile = path.join(program.directory, 'web/controllers/home.js')
    var exists = fs.existsSync(controllerFile)
    if (exists) return callback()
    var code = multiline.stripIndent(function() {/*
      exports.run = function(backbeam, req, res, libs, logger) {
        res.send('Hello world')
      }
    */})
    fs.writeFile(controllerFile, code, 'utf8', callback)
  })
  .then(function(callback) {
    var appjs = path.join(program.directory, 'app.js')
    var exists = fs.existsSync(appjs)
    if (exists) return callback()

    var code = multiline.stripIndent(function() {/*
      var http = require('http')
      var backbeam = require('backbeam-server')

      var app = backbeam.createExpressApp({ directory: __dirname })
      exports.app = app

      if (module.id === require.main.id) {
        var port = 3000
        http.createServer(app).listen(port)
        console.log('Started server at port %d', port)
      }
    */})
    fs.writeFile(appjs, code, 'utf8', callback)
  })
  .then(function(callback) {
    var packageJson = path.join(program.directory, 'package.json')
    var exists = fs.existsSync(packageJson)
    if (exists) return callback()

    var data = {
      name: projectName,
      version: '1.0.0',
      description: 'Node.js app made with Backbeam',
      scripts: {
        start: 'backbeam start',
        test: 'NODE_ENV=test mocha --bail --reporter spec test/',
      },
      author: {
        name: process.env.USER || '',
      },
      license: 'ISC',
      dependencies: {
        'backbeam-server': '^'+version,
      },
      devDependencies: {
        'mocha': '^1.21.4',
      },
      keywords: ['backbeam'],
    }
    fs.writeFile(packageJson, JSON.stringify(data, null, 2), 'utf8', callback)
  })
  .then(function(callback) {
    var testUtils = path.join(program.directory, 'test', 'test-utils.js')
    var exists = fs.existsSync(testUtils)
    if (exists) return callback()

    var code = multiline.stripIndent(function() {/*
      var txain = require('txain')
      var assert = require('assert')

      var request = require('supertest')(require('../app').app)
      exports.request = request

      exports.cleanDatabase = function(callback) {
        txain(function(callback) {
          request
            .post('/admin/delete-data')
            .end(function(err, res) {
              assert.ifError(err)
              assert.equal(200, res.statusCode, res.text)
              callback()
            })
        })
        .then(function(callback) {
          request
            .post('/admin/migrate')
            .end(function(err, res) {
              assert.ifError(err)
              assert.equal(200, res.statusCode, res.text)
              callback()
            })
        })
        .end(function(err) {
          assert.ifError(err)
          callback()
        })
      }
    */})
    fs.writeFile(testUtils, code, 'utf8', callback)
  })
  .end(callback)

  function randomToken(len) {
    return crypto.randomBytes(len).toString('hex')
  }

  var host = 'localhost'
  var emailTemplates = {
    registration: {
      subject: 'Welcome',
      txt: [
          'Hi there,',
          'We need to verify your email address.',
          'Just click here: http://'+host+'/common/verify/{{code}}',
        ].join('\n\n'),
      html: [
          'Hi there,<br><br>',
          'We need to verify your email address.<br><br>',
          'Just click <a href="http://'+host+'/common/verify/{{code}}">here</a>',
        ].join('\n\n'),
    },
    lostpassword: {
      subject: 'Reset your password',
      txt: [
          'Hi there,',
          'Did you forget your password? Ok, click here for instructions',
          'http://'+host+'/common/setpassword/{{code}}',
        ].join('\n\n'),
      html: [
          'Hi there,<br><br>',
          'Did you forget your password? Ok, click',
          '<a href="http://'+host+'/common/setpassword/{{code}}">here</a> for instructions',
        ].join('\n\n'),
    },
    confirm: {
      subject: 'Confirm your new email address',
      txt: [
          'Hi there,',
          'We need to verify your email address.',
          'Just click here: http://'+host+'/common/verify/{{code}}',
        ].join('\n\n'),
      html: [
          'Hi there,',
          '<br><br> We need to verify your email address.<br><br>',
          'Just click <a href="http://'+host+'/common/verify/{{code}}">here</a>',
      ].join('\n\n')
    }
  }
}

if (module.id === require.main.id) {
  exports.run()
}
